source/elements/oneMKL/source/domains/spblas/operations/spmv.rst

```diff
@@ -0,0 +1,371 @@
+
+       enum class spmv_alg {
+           default_alg,
+           coo_alg1,
+           coo_alg2,
+           csr_alg1,
+           csr_alg2,
+           csr_alg3,
+       };
+
+   }
+
+.. list-table::
+   :header-rows: 1
+   :widths: 10 30 45
+
+   * - Value
+     - Description
+     - Backend equivalent
+   * - ``default_alg``
+     - Default algorithm.
+     - | MKL: N/A
+       | cuSPARSE: ``CUSPARSE_SPMV_ALG_DEFAULT``
+       | rocSPARSE: ``rocsparse_spmv_alg_default``
+   * - ``coo_alg1``
+     - Default algorithm for COO format.
+     - | MKL: N/A
+       | cuSPARSE: ``CUSPARSE_SPMV_COO_ALG1``
+       | rocSPARSE: ``rocsparse_spmv_alg_coo``
+   * - ``coo_alg2``
+     - Deterministic algorithm for COO format.
+     - | MKL: N/A
+       | cuSPARSE: ``CUSPARSE_SPMV_COO_ALG2``
+       | rocSPARSE: ``rocsparse_spmv_alg_coo_atomic``
+   * - ``csr_alg1``
+     - Default algorithm for CSR format.
+     - | MKL: N/A
+       | cuSPARSE: ``CUSPARSE_SPMV_CSR_ALG1``
+       | rocSPARSE: ``rocsparse_spmv_alg_csr_adaptive``
+   * - ``csr_alg2``
+     - Deterministic algorithm for CSR format.
+     - | MKL: N/A
+       | cuSPARSE: ``CUSPARSE_SPMV_CSR_ALG2``
+       | rocSPARSE: ``rocsparse_spmv_alg_csr_stream``
+   * - ``csr_alg3``
+     - LRB variant of the algorithm for CSR format.
+     - | MKL: N/A
+       | cuSPARSE: N/A
+       | rocSPARSE: ``rocsparse_spmv_alg_csr_lrb``
+
+.. _onemkl_sparse_spmv:
+
+spmv
+----
+
+.. rubric:: Syntax
+
+.. code-block:: cpp
+
+   namespace oneapi::mkl::sparse {
+
+       void spmv_buffer_size(
+           sycl::queue                                &queue,
+           oneapi::mkl::transpose                     opA,
+           void*                                      alpha,
+           oneapi::mkl::sparse::matrix_view           A_view,
+           oneapi::mkl::sparse::matrix_handle_t       A_handle,
+           oneapi::mkl::sparse::dense_vector_handle_t x_handle,
+           void*                                      beta,
+           oneapi::mkl::sparse::dense_vector_handle_t y_handle,
+           oneapi::mkl::sparse::spmv_alg              alg,
+           oneapi::mkl::sparse::spmv_descr_t          spmv_descr,
+           std::size_t                                &temp_buffer_size,
+       );
+
+       void spmv_optimize(
+           sycl::queue                                &queue,
+           oneapi::mkl::transpose                     opA,
+           void*                                      alpha,
+           oneapi::mkl::sparse::matrix_view           A_view,
+           oneapi::mkl::sparse::matrix_handle_t       A_handle,
+           oneapi::mkl::sparse::dense_vector_handle_t x_handle,
+           void*                                      beta,
+           oneapi::mkl::sparse::dense_vector_handle_t y_handle,
+           oneapi::mkl::sparse::spmv_alg              alg,
+           oneapi::mkl::sparse::spmv_descr_t          spmv_descr,
+           sycl::buffer<std::uint8_t, 1>              workspace,
```

Comment:
I don't think there should be? Although the implementations of sub-buffers are probably not as much tested as buffers and USM pointers.

Reply:
For the implementation of the DFT external workspaces, there was a restriction on the use of sub-buffers in the Intel oneMKL API. My concern was that perhaps there is a similar restriction in a sparse API. I had a look at the docs (eg. [here](https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-dpcpp/2024-0/oneapi-mkl-sparse-matmat.html) and did not find any such restrictions.

Reply:
The workspace should not be a `sub_buffer` because oneMKL internally breaks it down into sub-buffers for its use.

Reply:
Thanks that makes sense. Do the other buffers also have this restriction?

Reply:
I don't see why we can't allow dense matrices or vectors to be sub-buffers. Likewise, I don't see why the sparse matrix arrays can't be sub-buffers too. At least in Intel oneMKL, we don't split the user-supplied dense matrices/vectors or sparse matrix arrays into sub-buffers. It's only the workspace that must not be a sub-buffer because the backend breaks it down into sub-buffers internally based on the choice of algorithm.

Reply:
@spencerpatty, do you have thoughts on this?

Reply:
we definitely do not need to specify that sparse matrix arrays or other data arrays(vectors/matrices) cannot be sub-buffers.  It is sufficient to say that the user-provided workspaces must not be sub-buffers as we may break it into sub-buffers internally, but all other data inputs are not disallowed from being sub-buffers.

Reply:
I added the restriction to prevent workspaces from being a sub-buffer in https://github.com/oneapi-src/oneAPI-spec/pull/522/commits/81b8be55f336b40be500d46c0809cb07bd0a5c34

The paragraph is split on multiple lines like so:
![workspace](https://github.com/oneapi-src/oneAPI-spec/assets/33024478/36fec76e-cfa5-4472-957c-fd5557df84fd)

Reply:
This looks good to me.